server/src/main/kotlin/suwayomi/tachidesk/graphql/mutations/SettingsMutation.kt

```diff
@@ -0,0 +1,89 @@
+        if (settings.webUIUpdateCheckInterval != null) serverConfig.webUIUpdateCheckInterval.value = settings.webUIUpdateCheckInterval!!
+
+        if (settings.downloadAsCbz != null) serverConfig.downloadAsCbz.value = settings.downloadAsCbz!!
+        if (settings.downloadsPath != null) serverConfig.downloadsPath.value = settings.downloadsPath!!
+        if (settings.autoDownloadNewChapters != null) serverConfig.autoDownloadNewChapters.value = settings.autoDownloadNewChapters!!
+
+        if (settings.maxSourcesInParallel != null) serverConfig.maxSourcesInParallel.value = settings.maxSourcesInParallel!!
+
+        if (settings.excludeUnreadChapters != null) serverConfig.excludeUnreadChapters.value = settings.excludeUnreadChapters!!
+        if (settings.excludeNotStarted != null) serverConfig.excludeNotStarted.value = settings.excludeNotStarted!!
+        if (settings.excludeCompleted != null) serverConfig.excludeCompleted.value = settings.excludeCompleted!!
+        if (settings.globalUpdateInterval != null) serverConfig.globalUpdateInterval.value = settings.globalUpdateInterval!!
+
+        if (settings.basicAuthEnabled != null) serverConfig.basicAuthEnabled.value = settings.basicAuthEnabled!!
+        if (settings.basicAuthUsername != null) serverConfig.basicAuthUsername.value = settings.basicAuthUsername!!
+        if (settings.basicAuthPassword != null) serverConfig.basicAuthPassword.value = settings.basicAuthPassword!!
+
+        if (settings.debugLogsEnabled != null) serverConfig.debugLogsEnabled.value = settings.debugLogsEnabled!!
+        if (settings.systemTrayEnabled != null) serverConfig.systemTrayEnabled.value = settings.systemTrayEnabled!!
+
+        if (settings.backupPath != null) serverConfig.backupPath.value = settings.backupPath!!
+        if (settings.backupTime != null) serverConfig.backupTime.value = settings.backupTime!!
+        if (settings.backupInterval != null) serverConfig.backupInterval.value = settings.backupInterval!!
+        if (settings.backupTTL != null) serverConfig.backupTTL.value = settings.backupTTL!!
+
+        if (settings.localSourcePath != null) serverConfig.localSourcePath.value = settings.localSourcePath!!
+    }
+
+    fun setSettings(input: SetSettingsInput): SetSettingsPayload {
+        val (clientMutationId, settings) = input
+
+        updateSettings(settings)
+
+        return SetSettingsPayload(clientMutationId, SettingsType())
+    }
+
+    data class ResetSettingsInput(val clientMutationId: String?)
+
+    data class ResetSettingsPayload(
+        val clientMutationId: String?,
+        val settings: SettingsType
+    )
+
+    fun resetSettings(input: ResetSettingsInput): ResetSettingsPayload {
+        val (clientMutationId) = input
+
+        GlobalConfigManager.resetUserConfig()
+        val defaultServerConfig = ServerConfig({ GlobalConfigManager.config.getConfig(SERVER_CONFIG_MODULE_NAME) })
```

Comment:
This resets the running server? All the MutableStateFlows update properly?

Reply:
yes, this is what happens:

 - `GlobalConfigManager.resetUserConfig()` resets the config file
 - `val defaultServerConfig = ...` creates a new ServerConfig instance to access the variables without having to manually retrieve them from the config
 - `updateSettings(settings)` updates the flows

and I have to add a guard to prevent updating the config file in case the value that should be set is the same as the current value in the config
since currently the file gets reset and then every flow value change updates the file again...

Reply:
added guard